### DISCUSSION TITLE
2 SessionProvider instances side-effect prevents session initial update()

### DISCUSSION THREAD

**Discussion by Thinkscape** [nextauthjs/next-auth#7324] [Help]
### Environment

```
    "next": "13.3.0",
    "next-auth": "^4.22.1",
    "react": "18.2.0",
```
```
  System:
    OS: macOS 13.3.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 186.88 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
  Browsers:
    Chrome: 112.0.5615.121
    Firefox: 104.0.2
    Safari: 16.4
```

### Reproduction URL

is there a simpler version of next-auth-example where I don't need to configure oauth2 providers?

### Describe the issue

There seems to be a side effect when using 2 `<NextAuthSessionProvider />` instances, one pre-populated, and other not.
This happens only on production next build (i.e. `next build && next start`).

By having two NextAuthSessionProvider instances, one being pre-populated, it causes the second one to refuse to load the initial session data, unless manually triggered with `update()`.

### How to reproduce

* latest `next-auth` and `next`
* a page, a session provider client component, and session consumer client component
* production build with `next build && next start` or deploy to Vercel
* open the following page in a browser

`somepage/page.tsx`
```tsx
export default async function SomePage() {
  const session = await getServerSession();

  return (
    <>
      <h3 className="m-2 rounded-md bg-slate-200 pl-2">
        Session pre-hydrated from RSC
      </h3>
      <SessionProvider session={session}>
        <ClientSessionConsumer />
      </SessionProvider>
      <h3 className="m-2 rounded-md bg-slate-200 pl-2">
        Client component without any data
      </h3>
      <SessionProvider>
        <ClientSessionConsumer />
      </SessionProvider>
    </>
  );
}
```
`SessionProvider.tsx:`
```tsx
"use client";

import { SessionProvider as NextAuthSessionProvider } from "next-auth/react";
import { type Session } from "next-auth";

export default function SessionProvider({
  children,
  session,
}: {
  children: React.ReactNode;
  session?: Session | null;
}) {
  return (
    <NextAuthSessionProvider session={session}>
      {children}
    </NextAuthSessionProvider>
  );
}
```
`ClientSessionConsumer.tsx`
```tsx
"use client";

import { useSession } from "next-auth/react";

export default function ClientSessionConsumer() {
 const session = useSession();
  return (
    <>
      <pre className="max-w-sm pl-2 font-mono text-xs md:pl-5">
        {JSON.stringify(session, null, 2)}
      </pre>
    </>
  );
}
```

## Expected

* The first component gets session data right away.
* The second component renders first without data, then loads from `/api/auth/session` and re-renders with session data

## Actual

* The first component gets session data right away.
* The second component flashes `status: "loading"` and then hangs at `status: "unauthenticated"`
* There are no network requests sent to `/api/auth/session` endpoint

```
{
  "data": null,
  "status": "unauthenticated"
}
```

**Comment by Thinkscape** [comment]
Just curious @balazsorban44, why do you believe it's _not_ a bug, and instead a discussion?

The Provider uses `React.context` under the hood, which itself is free of side-effects and allows having different context boundaries. `NextAuthSessionProvider` instances however seem to affect each other, which is IMHO either an undocumented shared state (cache? optimisation?) or a bug.

If using multiple `NextAuthSessionProvider` in one Next app is forbidden for some reason, then it should be documented.

**Reply by balazsorban44** [reply]
I don't think we ever documented multiple `SessionProvider`s, and most likely it would not be needed either. We recommend setting it up in `_app`, wrapping all the `useSession` hooks down the tree instead: https://next-auth.js.org/getting-started/client#sessionprovider

We are moving away from `SessionProvider` in general as the App Router is nearing stability, so I considered this an edge-case low priority in any case.

**Reply by Thinkscape** [reply]
Interesting. I'm using app router exclusively.

If you're moving away from `SessionProvider`, then how will client components access the session? (would they all require to be manually passed the session via primary props populated by RSC?)
